eclipse/src/saros/editor/internal/ContributionAnnotationManager.java

```diff
@@ -112,102 +101,151 @@ public ContributionAnnotationManager(
+   * {@link IDocument} that is performed by the caller of {@link #insertAnnotation} leads to an
+   * internal change of the annotation model.
+   *
+   * <p>The annotation with length &gt; 1 is removed, because the new annotation is subsequently
+   * added by {@link #insertAnnotation}
    *
-   * @param model to search for annotations to split
-   * @param offset at which annotations should be split
+   * <p>Therefore, this methods works against the internal mechanisms of eclipse in order to enforce
+   * the invariant of annotations with have a length &lt;= 1.
+   *
+   * @param model The model that could contain a violation of the invariant.
    */
-  public void splitAnnotation(final IAnnotationModel model, final int offset) {
-
-    if (!contribtionAnnotationsEnabled) return;
+  private void enforceAnnotationsWithLengthOne(IAnnotationModel model) {
+    List<ContributionAnnotation> annotationsToRemove = new ArrayList<>();
 
-    final List<ContributionAnnotation> annotationsToRemove =
-        new ArrayList<ContributionAnnotation>();
-
-    final Map<ContributionAnnotation, Position> annotationsToAdd =
-        new HashMap<ContributionAnnotation, Position>();
-
-    for (Iterator<Annotation> it = model.getAnnotationIterator(); it.hasNext(); ) {
-
-      final Annotation annotation = it.next();
+    Iterator<Annotation> i = model.getAnnotationIterator();
+    while (i.hasNext()) {
+      Annotation annotation = i.next();
 
       if (!(annotation instanceof ContributionAnnotation)) continue;
 
-      final ContributionAnnotation contributionAnnotation = (ContributionAnnotation) annotation;
-      final User source = contributionAnnotation.getSource();
-
-      final Position pos = model.getPosition(contributionAnnotation);
+      ContributionAnnotation contrAnnotation = (ContributionAnnotation) annotation;
+      Position p = model.getPosition(contrAnnotation);
+      if (p.getLength() > 1) {
+        annotationsToRemove.add(contrAnnotation);
 
-      if ((offset > pos.offset) && (offset < pos.offset + pos.length)) {
+        // Expect only one annotation with length &gt; 1 created by
+        // callers content change.
+        break;
+      }
+    }
+    annotationModelHelper.replaceAnnotationsInModel(
+        model, annotationsToRemove, Collections.emptyMap());
+  }
 
-        final Position before = new Position(pos.offset, offset - pos.offset);
+  /**
+   * Creates contribution annotations with length 1 for each char contained in the range defined by
+   * {@code offset} and {@code length}.
+   *
+   * @param model the model that is assigned to the created contribution annotations.
+   * @param offset start of the annotation range.
+   * @param length length of the annotation range.
+   * @param source of the annotation.
+   * @return a map containing the annotations and their positions
+   */
+  private Map<ContributionAnnotation, Position> createAnnotationsForContributionRange(
```

Comment:
So the annotations are present twice in this case, meaning you can just remove the merged one?

Is there currently any logic re-adding existing annotations to the model? If so, can this become an issue in such cases as well (i.e. the internal optimization to merge the annotations is applied again)?

In general, does this have a noticeable impact on performance?

Reply:
I found this optimization only if the IDocument#replace method is used. Therefore, I assume that this is not an optimization of the model, but the way the replace method introduces annotations. I also tested the re-adding (e.g. if the use color changes).

With using the optimized replaceAnnotations method I noticed no impact, without the optimized version you notice an impact if you paste long text (the annotations are painted bit by bit). Therefore, we use the optimized version.